### PR TITLE
made ObjectWriter/Reader to be able to write primitive type array

### PR DESCRIPTION
**Pull request by heejaechang**

### CODE REVIEW

**Comment by heejaechang** [comment]
@mattwar @dotnet/roslyn-ide @mavasani can you take a look?

File: src/Compilers/Core/Portable/Serialization/ObjectReader.cs [hunk 1]
```diff
@@ -334,24 +334,102 @@ private Array ReadArray(DataKind kind)
...
+                    case DataKind.Float4:
+                        array.SetValue(_reader.ReadSingle(), i);
+                        continue;
+                    case DataKind.Float8:
+                        array.SetValue(_reader.ReadDouble(), i);
+                        continue;
+                    case DataKind.Decimal:
+                        array.SetValue(_reader.ReadDecimal(), i);
+                        continue;
+                    case DataKind.BooleanType:
```
Comment (mattwar) [inline src/Compilers/Core/Portable/Serialization/ObjectReader.cs]: Opportunity here for array of booleans to be packed bits.  Just saying.. :-)

**Comment by mattwar** [comment]
:+1:

File: src/Compilers/Core/Portable/Serialization/ObjectReader.cs (hunk 1 above)
Comment (heejaechang) [inline src/Compilers/Core/Portable/Serialization/ObjectReader.cs]: will do that in next checkin.

File: src/Compilers/Core/Portable/Serialization/ObjectReader.cs [hunk 2]
```diff
@@ -334,24 +334,102 @@ private Array ReadArray(DataKind kind)
...
+
+            // optimization for primitive type array
+            Type elementType;
+            if (s_reverseTypeMap.TryGetValue(elementKind, out elementType))
+            {
+                return this.ReadPrimitiveTypeArrayElements(elementType, elementKind, length);
+            }
+
+            // custom type case
+            elementType = this.ReadType(elementKind);
```
Comment (tmat) [inline src/Compilers/Core/Portable/Serialization/ObjectReader.cs]: Wouldn't it be more efficient to do something like:

```
switch (kind)
{
   case DataKind.Int8: 
      array = new sbyte[length];
      for (int i = 0; i < array.Length; i++) array[i] = ReadSByte();
      break;

   case DataKind.UInt8: 
      array = new byte[length];
      for (int i = 0; i < array.Length; i++) array[i] = ReadByte();
      break;
...
}
```

That is, specialize the entire array creation and initialization by element kind?

Comment (heejaechang) [inline src/Compilers/Core/Portable/Serialization/ObjectReader.cs]: sure I can do that.

**Pull request merged**